### PR TITLE
Add response id to request error. (bunq/sdk_php#88)

### PR DESCRIPTION
**Pull request by OGKevin**
Closes bunq/sdk_php#88
 - [x] Tested

### CODE REVIEW

**Comment by OGKevin** [comment]
@sandervdo please 👀

File: src/Exception/ExceptionFactory.php [hunk 1]
```diff
@@ -30,45 +32,57 @@ class ExceptionFactory
...
      * @param string[] $messages
+     * @param string $responseId
      *
      * @return string
      */
-    private static function generateErrorMessage(int $responseCode, array $messages): string
-    {
+    private static function generateErrorMessage(
+        int $responseCode,
+        array $messages,
```
Comment (sandervdo) [inline src/Exception/ExceptionFactory.php]: allMessage

File: src/Exception/ExceptionFactory.php [hunk 2]
```diff
@@ -30,45 +32,57 @@ class ExceptionFactory
...
      * @param int $responseCode
+     * @param string $responseId
      *
      * @return ApiException
      */
-    public static function createExceptionForResponse(array $messages, int $responseCode): ApiException
-    {
-        $errorMessage = static::generateErrorMessage($responseCode, $messages);
+    public static function createExceptionForResponse(
+        array $messages,
```
Comment (sandervdo) [inline src/Exception/ExceptionFactory.php]: allMessage

File: src/Exception/ExceptionFactory.php [hunk 3]
```diff
@@ -20,6 +20,8 @@ class ExceptionFactory
      * Formatting constants
      */
     const FORMAT_RESPONSE_CODE_LINE = 'HTTP Response Code: %s';
+    const FORMAT_RESPONSE_ID = 'The response id to help bunq debug: %s';
+    const FORMAT_ERROR_MESSAGE_LINE = 'Error message: %s';
     const GLUE_ERROR_MESSAGES = "\n";
```
Comment (sandervdo) [inline src/Exception/ExceptionFactory.php]: ERROR_MESSAGE_GLUE? ERROR_MESSAGE_SEPARATOR might even be better

File: src/Http/Handler/ResponseHandlerError.php [hunk 4]
```diff
@@ -3,24 +3,38 @@
...
     const STATUS_CODE_OK = 200;
 
+    /**
+     * The index of the first item in an array.
+     */
+    const INDEX_FIRST = 0;
+
     /**
      * @param ResponseInterface $response
      *
```
Comment (sandervdo) [inline src/Http/Handler/ResponseHandlerError.php]: `@return`

File: src/Http/Handler/ResponseHandlerError.php [hunk 5]
```diff
@@ -65,11 +80,32 @@ private function fetchErrorDescriptions(array $errorArray): array
     {
         $errorDescriptions = [];
```
Comment (sandervdo) [inline src/Http/Handler/ResponseHandlerError.php]: allErrorDescription

File: src/Http/Handler/ResponseHandlerError.php [hunk 6]
```diff
@@ -49,9 +64,9 @@ private function fetchErrorMessages(ResponseInterface $response): array
         $responseBody = $response->getBody();
         $responseBodyInJson = json_decode($responseBody, true);
 
-        if ($responseBodyInJson !== false){
+        if ($responseBodyInJson !== false) {
             return $this->fetchErrorDescriptions($responseBodyInJson);
```
Comment (sandervdo) [inline src/Http/Handler/ResponseHandlerError.php]: fetchAllErrorDescription

File: tests/Http/ErrorResponseIdTest.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<?php
+
```
Comment (sandervdo) [inline tests/Http/ErrorResponseIdTest.php]: remove empty line

File: tests/Http/ErrorResponseIdTest.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Tests if the response id from a failed request can be retrieved successfully.
+ */
+class ErrorResponseIdTest extends BunqSdkTestBase
+{
+    /**
+     */
+    public function testBadRequestWitResponseId()
+    {
+        $caughtExcretion = null;
```
Comment (sandervdo) [inline tests/Http/ErrorResponseIdTest.php]: Did you mean exception?

File: tests/Http/ErrorResponseIdTest.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+        try {
+            UserPerson::get(static::$apiContext, 0);
```
Comment (sandervdo) [inline tests/Http/ErrorResponseIdTest.php]: Magic number

File: src/Http/Handler/ResponseHandlerError.php (hunk 4 above)
Comment (OGKevin) [reply src/Http/Handler/ResponseHandlerError.php]: Invalid, as you can see this particular line has not been changed in this pr and therefore only a snippet of the doc block is show. If you look at the original file you see that this doc block is correct 🙃 see: https://github.com/bunq/sdk_php/blob/1a5188bcf53e836912ae01a82c585b89219a075f/src/Http/Handler/ResponseHandlerError.php#L41

Comment (sandervdo) [reply src/Http/Handler/ResponseHandlerError.php]: Ah yes, missed that one. Ignore then. 😄

**Comment by OGKevin** [comment]
@sandervdo all yours again, please 👀

File: src/Exception/ExceptionFactory.php [hunk 10]
```diff
@@ -20,55 +20,73 @@ class ExceptionFactory
...
+    ): string {
         $lineResponseCode = sprintf(self::FORMAT_RESPONSE_CODE_LINE, $responseCode);
+        $lineResponseId = sprintf(self::FORMAT_RESPONSE_ID, $responseId);
+        $lineErrorMessage = sprintf(
+            self::FORMAT_ERROR_MESSAGE_LINE,
+            implode($allMessage)
+        );
 
-        return static::glueMessages(array_merge([$lineResponseCode], $messages));
+        return static::glueMessages([$lineResponseCode, $lineResponseId, $lineErrorMessage]);
```
Comment (sandervdo) [inline src/Exception/ExceptionFactory.php]: Oh crap, I see a plural here now too. 🙈

File: src/Exception/ExceptionFactory.php [hunk 11]
```diff
@@ -78,6 +96,6 @@ private static function generateErrorMessage(int $responseCode, array $messages)
      */
     private static function glueMessages(array $messages): string
     {
-        return implode(self::GLUE_ERROR_MESSAGES, $messages);
+        return implode(self::SEPARATOR_ERROR_MESSAGES, $messages);
```
Comment (sandervdo) [inline src/Exception/ExceptionFactory.php]: Would call it `ERROR_MESSAGE_SEPARATOR` then, (big -> small)

Comment (OGKevin) [reply src/Exception/ExceptionFactory.php]: We normally group constants to getter, so in this case this constant belongs to the `SEPARATOR_` constant group. As you can have multiple separators and then the constant names will begin with `SEPARATOR_XXXXXX`.

**Comment by OGKevin** [comment]
@sandervdo all yours again, please 👀

**Comment by OGKevin** [comment]
@andrederoos All yours.

**Comment by andrederoos** [comment]
@OGKevin branche is out-of-date

**Comment by OGKevin** [comment]
@andrederoos my bad, updated 🤦‍♂️

**Pull request merged**